### ISSUE TITLE
[FEATURE] Faster ray-triangle intersection test that does not allocate large arrays

### ISSUE THREAD

**Issue by jeertmans** [jeertmans/DiffeRT#313]
### Terms

- [x] Checked the [existing issues](https://github.com/jeertmans/DiffeRT/issues?q=is%3Aissue+label%3Aenhancement) to see if my suggestion has not already been suggested;

### Description

The ray-intersect-any-triangle test is one of the main performance bottlenecks in this library when applied to a large number of rays and triangles. This is because `jax.jit` cannot optimize the `jax.vmap`-ed version of this check, resulting in the allocation of an array whose size scales with `num_rays x num_triangles`-too large to fit into any reasonable amount of memory.

To work around this, the current implementation uses a `jax.lax.scan`-like approach that processes triangles (or rays) sequentially, in user-defined batch sizes. The main drawbacks are:

* The optimal batch size depends on many factors, including the number of rays, the number of triangles, and the available memory, making it difficult to choose good default values.
* `jax.lax.scan`-like solutions are orders of magnitude slower than their `jax.vmap` counterparts when array sizes are not prohibitively large.

Ideally, we would develop a `jax.vmap` solution that `jax.jit` can optimize in a way that avoids allocating the problematic intermediate array altogether. I have already spent quite some time investigating this, but so far, without success.

If you have ideas for improving this or would like to contribute, please share your thoughts or open a PR!

### Additional information

**Related links:** 
- My discussion opened in JAX's repo: https://github.com/jax-ml/jax/discussions/30470
- My issue opened in JAX's repo: https://github.com/jax-ml/jax/issues/30841
- Example issue where JAX can optimize the reduction: https://github.com/jax-ml/jax/discussions/9505
- JAX's [`Ref`](https://docs.jax.dev/en/latest/_autosummary/jax.ref.Ref.html) could be interesting to look at: https://docs.jax.dev/en/latest/array_refs.html

**Relevant functions:**
- [`rays_intersect_any_triangle`](https://differt.eertmans.be/latest/reference/_autosummary/differt.rt.rays_intersect_any_triangle.html#differt.rt.rays_intersect_any_triangle);
- [`triangles_visible_from_vertices`](https://differt.eertmans.be/latest/reference/_autosummary/differt.rt.triangles_visible_from_vertices.html#differt.rt.triangles_visible_from_vertices);
- [`first_triangles_hit_by_rays`](https://differt.eertmans.be/latest/reference/_autosummary/differt.rt.first_triangles_hit_by_rays.html#differt.rt.first_triangles_hit_by_rays).
